password-instances/src/Data/Password/Instances.hs

```diff
@@ -14,90 +14,79 @@ Maintainer  : cdep.illabout@gmail.com
--- >>> fmap unsafeShowPassword maybePass
+-- >>> let maybePassword = decode "\"foobar\"" :: Maybe Password
+-- >>> fmap unsafeShowPassword maybePassword
 -- Just "foobar"
 --
--- There is no instance for 'ToJSON' for 'Pass' because we don't want to
--- accidentally encode a plain-text 'Pass' to JSON and send it to the end-user.
+-- There is no instance for 'ToJSON' for 'Password' because we don't want to
+-- accidentally encode a plain-text 'Password' to JSON and send it to the end-user.
 --
--- Similarly, there is no 'ToJSON' and 'FromJSON' instance for 'PassHash'
+-- Similarly, there is no 'ToJSON' and 'FromJSON' instance for 'PasswordHash'
 -- because we don't want to accidentally send the password hash to the end
 -- user.
-instance FromJSON Pass where
-  parseJSON = fmap mkPass . parseJSON
+instance FromJSON Password where
+  parseJSON = fmap mkPassword . parseJSON
 
--- | This instance allows a 'Pass' to be created with functions like
+-- | This instance allows a 'Password' to be created with functions like
 -- 'Web.HttpApiData.parseUrlPiece' or 'Web.HttpApiData.parseQueryParam'.
 --
--- >>> let eitherPass = parseUrlPiece "foobar"
--- >>> fmap unsafeShowPassword eitherPass
+-- >>> let eitherPassword = parseUrlPiece "foobar"
+-- >>> fmap unsafeShowPassword eitherPassword
 -- Right "foobar"
-instance FromHttpApiData Pass where
-  parseUrlPiece = fmap mkPass . parseUrlPiece
+instance FromHttpApiData Password where
+  parseUrlPiece = fmap mkPassword . parseUrlPiece
 
--- | This instance allows a 'PassHash' to be stored as a field in a database using
+-- | This instance allows a 'PasswordHash' to be stored as a field in a database using
 -- "Database.Persist".
 --
 -- >>> let salt = Salt "abcdefghijklmnopqrstuvwxyz012345"
--- >>> let pass = mkPass "foobar"
--- >>> let hashedPassword = hashPassWithSalt salt pass
+-- >>> let pass = mkPassword "foobar"
+-- >>> let hashedPassword = hashPasswordWithSalt defaultParams salt pass
 -- >>> toPersistValue hashedPassword
--- PersistText "14|8|1|YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU=|nENDaqWBmPKapAqQ3//H0iBImweGjoTqn5SvBS8Mc9FPFbzq6w65maYPZaO+SPamVZRXQjARQ8Y+5rhuDhjIhw=="
+-- PersistText "16|8|1|YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU=|BH0oidcU/4Ec7Co4EM+LQ6xp39//MnOUhqmNeOnOz/nl4JHNXEJBw5dPdi3wTStYr+e1SmJkzHJrMvUJYNxK1w=="
 --
 -- In the example above, the long 'PersistText' will be the value you store in
 -- the database.
 --
--- We don't provide an instance of 'PersistField' for 'Pass', because we don't
+-- We don't provide an instance of 'PersistField' for 'Password', because we don't
 -- want to make it easy to store a plain-text password in the database.
-deriving newtype instance PersistField PassHash
+instance PersistField (PasswordHash a) where
+  toPersistValue (PasswordHash hpw) = PersistText hpw
+  fromPersistValue (PersistText txt) = Right $ PasswordHash txt
+  fromPersistValue (PersistByteString bs) = Right $ PasswordHash $ TE.decodeUtf8 bs
```

Comment:
I don't think it is a big issue, but we could probably take this line out?

```suggestion
```

If we decide to leave this line in, we should probably mention it in the documentation.

Reply:
I would not take this out. If someone used ByteString (SQL type: `bytea`) as type of the column for the hashes before using this package (or maybe even before using Haskell) it will fail to parse, which is unnecessary.

Reply:
Okay, that's a good point.  I agree with keeping it in then.